Update the golangci-lint.yaml

raghavendra-talur opened the pull request:
This is in preparation to updating the golangci-lint version. This PR is a no-op.

Just moving the deprecated linters to a new sorted section.

nirs commented on .golangci.yaml:
```diff
@@ -109,30 +109,37 @@ linters:
   disable:
+  # Keep this list sort and grouped by reason
+  # You can get the list by running the command
+  # golangci-lint linters | awk 'f;/Disabled by your/{f=1}' | cut -d":" -f1 | grep -v deprecated | cut -d" " -f1 | sort
```
This one liner is to complicated to be helpful. What are the too groups we want to separate?

Seems simpler just to sort the entire list and add a comment for some linters.

raghavendra-talur replied:
I felt this is easy to parse the configuration file in one look. Interspersed list is complicated to look at.

nirs replied:
Adding the complex commands and creating 2 groups of items in the same list is very messy. The current form having comment per line looks better to me.

nirs commented on .golangci.yaml:
```diff
@@ -140,6 +132,13 @@ linters:
     - nosnakecase
+    - scopelint
+    - structcheck
+    - varcheck
```
Why keep deprecated list instead of deleting them?

raghavendra-talur replied:
Just because we had enabled them at some point and I don't want the information to be lost. We can certainly delete them later once we agree on it.

nirs replied:
This change actually disable more linters, right? the diff is not clear because the lists are so long, and the commit message you don't list the disabled linters.

I support disabling the deprecated linters but make it clear in the commit message.

raghavendra-talur commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -653,7 +653,7 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
 		err2 := r.updateDRPCStatus(ctx, drpc, placementObj, logger)
 
-		return ctrl.Result{}, fmt.Errorf("failed to create DRPC instance (%w) and (%v)", err, err2)
+		return ctrl.Result{}, fmt.Errorf("failed to create DRPC instance (%w) and (%w)", err, err2)
```
This blog does a pretty good job of explaining why we have %v before and why %w is better. Basically, we got the capability on recently in go to join two errors.

Thanks for flagging this though. We cannot merge this patch until we update to go 1.20.

nirs replied:
We can separate this to another pr and wait until we can require 1.20.

raghavendra-talur commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -925,7 +925,7 @@ func (r *DRPlacementControlReconciler) getDRPolicy(ctx context.Context,
 	drpc *rmn.DRPlacementControl, placementObj client.Object, log logr.Logger,
 ) error {
-	update := false
+	var update bool
```
I was in two minds when I made this change. I don't like the new code.

However, the linter in question, wastedassign might be useful to catch a logical bug if one forgets to use a value before reusing the variable. I wasn't able to find a configuration for the linter that allows for the initialization especially to the null values of the types.

nirs replied:
This is very minor issue - if this linter is useful in other cases lets use it.

ShyamsundarR commented:
Older instance where a bunch of linters were disabled and we never go to rectifying the situation is [here](https://github.com/RamenDR/ramen/pull/534/files#diff-9917ddc9f1c3304218f7269265b746d997c5c0615478177b5fceecd33ef47cb5). Pointing this out to see how best to recover from that situation as well.

raghavendra-talur commented:
@nirs and @ShyamsundarR Thank you for your valuable feedback in this PR. I have addressed the comments and also split up the work in multiple PRs as this PR became too big with the changes. The initial work is merged in #1142 and #1143.

I will raise more PRs to address the issues that new linters or updated linters have brought up. I will also raise issues for the remaining linters so that others can pick up the work.
